src/Network/GRPC/MQTT/Sequenced.hs

```diff
@@ -4,146 +4,171 @@
-      readMessage acc = do
-        (Packet isLastPacket _ chunk) <- ExceptT readSeq
-        let builder = acc <> Builder.byteString chunk
-        if isLastPacket
-          then pure $ Builder.toLazyByteString builder
-          else readMessage builder
-
-  return $ readMessage mempty
-
-{- | Given an 'STM' action that gets a 'Sequenced' object, creates a new wrapped action that will
- read the objects in order even if they are received out of order
- NB: Objects with negative sequence numbers are always returned immediately
--}
-mkSequencedRead :: forall io a. (MonadIO io, Sequenced a) => STM a -> io (io a)
-mkSequencedRead read = do
-  seqVar <- newTVarIO 0
-  bufferVar <- newTVarIO $ SL.toSortedList @(SequencedWrap a) []
-
-  let orderedRead :: STM a
-      orderedRead = do
-        curSeq <- readTVar seqVar
-        buffer <- readTVar bufferVar
-
-        sa <- case SL.uncons buffer of
-          Just (sa, rest)
-            | SequencedIdx idx <- seqNum sa
-              , curSeq == idx ->
-              writeTVar bufferVar rest $> sa
-          _ -> coerce read
-
-        case seqNum sa of
-          Unsequenced -> return $ coerce sa
-          SequencedIdx idx ->
-            case idx `compare` curSeq of
-              -- If sequence number is less than current, it must be a duplicate, so we discard it and read again
-              LT -> orderedRead
-              EQ -> modifyTVar' seqVar (+ 1) $> coerce sa
-              GT -> modifyTVar' bufferVar (SL.insert sa) >> orderedRead
-
-  return $ atomically orderedRead
-
-mkStreamRead :: forall r io. (Message r, MonadIO io) => ExceptT RemoteError IO LByteString -> io (ExceptT RemoteError IO (Maybe r))
-mkStreamRead readRequest = do
-  reqsRef :: IORef (Maybe (Vector r)) <- newIORef (Just Vector.empty)
+
+--------------------------------------------------------------------------------
+
+-- | Reads, parses, and orders serialized 'Packet' messages from the source
+-- 'TChan'. Returns the serialized message reconstructed from each packet
+-- obtained from the channel.
+--
+-- @since 0.1.0.0
+packetReader ::
+  TChan Lazy.ByteString ->
+  ExceptT Decode.ParseError IO Lazy.ByteString
+packetReader channel = do
+  payloads <- liftIO (atomically (orderPacketReader channel))
+  case payloads of
+    Left err -> throwError err
+    Right xs -> pure (ByteString.Builder.toLazyByteString (builder xs))
+  where
+    builder :: [ByteString] -> ByteString.Builder
+    builder = foldl' (\xs x -> xs <> ByteString.Builder.byteString x) mempty
+
+-- | Reads a complete sequence of serialized 'Packet' messages from the source
+-- 'TChan' given. Returns an 'STM' action of sequenced packets read from the
+-- 'TChan' or a 'Decode.ParseError' raised while attempt to parse one or more
+-- of the channel's 'ByteString' as a 'Packet'.
+--
+-- @since 0.1.0.0
+orderPacketReader ::
+  TChan Lazy.ByteString ->
+  STM (Either Decode.ParseError [ByteString])
+orderPacketReader channel = do
+  buffer <- TMap.empty
```

Comment:
Does this need to use a `TMap`? The map is only used locally so all the thread-safe properties seem unnecessary.